src/diff/component.js

```diff
@@ -7,12 +7,132 @@ import {
 	DIRTY_BIT,
 	FORCE_UPDATE,
 	MODE_PENDING_ERROR,
-	MODE_RERENDERING_ERROR,
-	TYPE_CLASS
+	MODE_RERENDERING_ERROR
 } from '../constants';
 import { addCommitCallback } from './commit';
 import { getParentContext } from '../tree';
 
+export function renderFunctionComponent(
+	parentDom,
+	newVNode,
+	internal,
+	commitQueue,
+	startDom
+) {
+	/** @type {import('../internal').Component} */
+	let c;
+	let tmp, isNew;
+
+	/** @type {import('../internal').ComponentType} */
+	let type = (internal.type);
+
+	// @TODO split update + mount?
+	let newProps = newVNode ? newVNode.props : internal.props;
+
+	if (internal.flags & MODE_PENDING_ERROR) {
+		// Toggle the MODE_PENDING_ERROR and MODE_RERENDERING_ERROR flags. In
+		// actuality, this should turn off the MODE_PENDING_ERROR flag and turn on
+		// the MODE_RERENDERING_ERROR flag.
+		internal.flags ^= MODE_PENDING_ERROR | MODE_RERENDERING_ERROR;
+	}
+
+	const context = getParentContext(internal);
+
+	// Necessary for createContext api. Setting this property will pass
+	// the context value as `this.context` just for this component.
+	tmp = type.contextType;
+	let provider = tmp && context[tmp._id];
+	let componentContext = tmp
+		? provider
+			? provider.props.value
+			: tmp._defaultValue
+		: context;
+
+	if (internal && internal._component) {
+		c = internal._component;
+	} else {
+		internal._component = c = new Component(newProps, componentContext);
+
+		if (provider) provider._subs.add(internal);
+
+		c.props = newProps;
+		c.context = componentContext;
+		c.constructor = type;
+		c.render = doRender;
+		internal.flags |= DIRTY_BIT;
+		isNew = true;
+	}
```

Comment:
In an ideal world we get rid of `c` entirely but there's a few moving parts still relying on this like context